src/dfx/src/commands/deps/deploy.rs

```diff
@@ -0,0 +1,105 @@
+    get_pull_canisters_in_config, get_pulled_wasm_path, load_init_json, load_pulled_json,
+    validate_pulled,
+};
+use crate::lib::environment::Environment;
+use crate::lib::error::DfxResult;
+use crate::lib::root_key::fetch_root_key_if_needed;
+
+use anyhow::{anyhow, Context};
+use candid::Principal;
+use clap::Parser;
+use fn_error_context::context;
+use ic_agent::Agent;
+use ic_utils::interfaces::{management_canister::builders::InstallMode, ManagementCanister};
+use slog::{info, Logger};
+
+/// Deploy pulled canisters.
+#[derive(Parser)]
+pub struct DepsDeployOpts {
+    /// Specify the canister to deploy. You can specify its name (as defined in dfx.json) or Principal.
+    /// If not specified, all pulled canisters will be deployed.
+    canister: Option<String>,
+}
+
+pub async fn exec(env: &dyn Environment, opts: DepsDeployOpts) -> DfxResult {
+    let logger = env.get_logger();
+    let project_root = env.get_config_or_anyhow()?.get_project_root().to_path_buf();
+    let pulled_json = load_pulled_json(&project_root)?;
+    let pull_canisters_in_config = get_pull_canisters_in_config(env)?;
+    validate_pulled(&pulled_json, &pull_canisters_in_config)?;
+
+    let init_json = load_init_json(&project_root)?;
+
+    fetch_root_key_if_needed(env).await?;
+    let agent = env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+
+    match opts.canister {
+        Some(canister) => {
+            let canister_id = match pull_canisters_in_config.get(&canister) {
+                Some(canister_id) => *canister_id,
+                None => Principal::from_text(canister).with_context(|| {
+                    "The canister is not a valid Principal nor a name specified in dfx.json"
+                })?,
+            };
+            create_and_install(agent, logger, &canister_id, &init_json).await?;
+        }
+        None => {
+            for canister_id in pulled_json.canisters.keys() {
+                create_and_install(agent, logger, canister_id, &init_json).await?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+#[context("Failed to create and install canster {}", canister_id)]
+async fn create_and_install(
+    agent: &Agent,
+    logger: &Logger,
+    canister_id: &Principal,
+    init_json: &InitJson,
+) -> DfxResult {
+    let arg_raw = init_json.get_arg_raw(canister_id)?;
+    try_create_canister(agent, logger, canister_id).await?;
+    install_pulled_canister(agent, logger, canister_id, arg_raw).await?;
+    Ok(())
+}
+
+// not use operations::canister::create_canister because we don't want to modify canister_id_store
+#[context("Failed to create canster {}", canister_id)]
+async fn try_create_canister(agent: &Agent, logger: &Logger, canister_id: &Principal) -> DfxResult {
+    info!(logger, "Creating canister: {canister_id}");
+    let mgr = ManagementCanister::create(agent);
+    // ignore the error that the canister is already installed
+    let _res = mgr
```

Comment:
Is "canister is already installed" really the only possible error here?

Reply:
Of course there can be other possible error. I'm not sure if there are better approach to handle them.
If other error happens, the following "canister install" will fail and give some information there.

Reply:
Suggestion: add a method like this to state_tree/canister_info.rs and call it

```
pub async fn read_state_tree_canister_exists(
    agent: &Agent,
    canister_id: Principal,
) -> DfxResult<bool> {
    matches!(read_state_tree_canister_controllers(agent, canister_id).await, Ok(Some(_)))
}
```